src/VisualStudio/Core/Def/Implementation/Diagnostics/VisualStudioWorkspaceDiagnosticAnalyzerProviderService.cs

```diff
@@ -29,16 +31,39 @@ public VisualStudioWorkspaceDiagnosticAnalyzerProviderService(VisualStudioWorksp
             // Get the analyzer assets for installed VSIX extensions through the VSIX extension manager.
             var extensionManager = workspace.GetVsService<SVsExtensionManager, IVsExtensionManager>();
 
-            var builder = ImmutableArray.CreateBuilder<HostDiagnosticAnalyzerPackage>();
-            foreach (var extension in extensionManager.GetEnabledExtensions(AnalyzerContentTypeName))
+            // get rootfolder and shellfolder location
+            string rootFolder;
+            string shellFolder;
+            if (TryGetRootAndShellFolder(extensionManager, out shellFolder, out rootFolder))
```

Comment:
This looks really scary to me, do we really need to clone all of this VSExtensionManager's code?

Another approach would be to:
(a) Get all enabled extensions upfront and for each extension's content items (http://index/Microsoft.VisualStudio.ExtensionManager/R/8a2f139b4c502be6.html) just get the RelativePath (http://index/Microsoft.VisualStudio.ExtensionManager/R/d11f9db22ce337cf.html) and strip off leading chars to get just the content name (e.g. Microsoft.CodeAnalysis.CSharp.dll). Create a map from content name to containing extension name.

(b) Use the original code to invoke GetEnabledExtensionContentLocations and for each content location try to map content name part to the extension name using the above map.